File: 141. Linked List Cycle/c/linked_list_cycle.c
#include <stdio.h>
#include <stdlib.h>
// #include <assert.h>

// Definition for singly-linked list.
struct ListNode {
    int val;
    struct ListNode *next;
};

bool hasCycle(struct ListNode *head) {
    struct ListNode*fast = head;
    struct ListNode*slow = head;
    while (fast && fast->next) {
        if (slow == fast->next || slow->next == fast->next->next) {
            return 0;
        }
        fast = fast->next->next;
        slow = slow->next;
    }
    return 1;
}

File: 141. Linked List Cycle/go/linked_list_cycle.go
package linked_list_cycle

// Definition for singly-linked list.
type ListNode struct {
	Val int
	Next *ListNode
}

func hasCycle(head *ListNode) bool {
	slow := head
	fast := head
	for fast != nil && fast.Next != nil {
		if slow == fast.Next || slow.Next == fast.Next.Next {
			return true
		}
		slow = slow.Next
		fast = fast.Next.Next
		}
	return false
}